xrtl/base/arena_test.cc

```diff
@@ -0,0 +1,147 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "xrtl/base/arena.h"
+
+#include "xrtl/testing/gtest.h"
+
+namespace xrtl {
+namespace {
+
+// Tests basic block allocations.
+TEST(ArenaTest, BasicAllocation) {
+  Arena arena(64);
+  EXPECT_EQ(64, arena.block_size());
+  EXPECT_EQ(0, arena.bytes_allocated());
+  EXPECT_EQ(0, arena.block_bytes_allocated());
+
+  // Zero byte allocations should return nullptr and not allocate bytes.
+  auto zero_ptr = reinterpret_cast<uintptr_t>(arena.AllocateBytes(0));
+  EXPECT_EQ(0, zero_ptr);
+  EXPECT_EQ(0, arena.bytes_allocated());
+  EXPECT_EQ(0, arena.block_bytes_allocated());
+
+  arena.Clear();
+
+  // Allocations must be machine word aligned.
+  auto one_ptr = reinterpret_cast<uintptr_t>(arena.AllocateBytes(1));
+  EXPECT_NE(0, one_ptr);
+  EXPECT_EQ(0, one_ptr % sizeof(uintptr_t));
+  one_ptr = reinterpret_cast<uintptr_t>(arena.AllocateBytes(1));
+  EXPECT_NE(0, one_ptr);
+  EXPECT_EQ(0, one_ptr % sizeof(uintptr_t));
+  EXPECT_EQ(2, arena.bytes_allocated());
+  EXPECT_LT(2, arena.block_bytes_allocated());
+
+  arena.Clear();
+  EXPECT_EQ(0, arena.bytes_allocated());
+  EXPECT_EQ(0, arena.block_bytes_allocated());
+}
+
+// Tests typed allocations.
+TEST(ArenaTest, TypedAllocations) {
+  Arena arena(64);
+
+  EXPECT_NE(nullptr, arena.Allocate<int>());
+  EXPECT_EQ(4, arena.bytes_allocated());
+  EXPECT_EQ(80, arena.block_bytes_allocated());
```

Comment:
chunk size of 64 passed to the arena + some overhead. changed this to use a constant Arena::kBlockOverhead